nodebuilder/tests/p2p_test.go

```diff
@@ -86,131 +95,123 @@ Steps:
  5. Ensure that nodes are connected to bridge
  6. Wait until light will find full node
  7. Check that full and light nodes are connected to each other
- 8. Stop FN and ensure that it's not connected to LN
 */
-func TestBootstrapNodesFromBridgeNode(t *testing.T) {
+func TestFullDiscoveryViaBootstrapper(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), swamp.DefaultTestTimeout)
+	t.Cleanup(cancel)
+
+	const defaultTimeInterval = time.Second * 2
+
 	sw := swamp.NewSwamp(t)
+
+	// create and start a BN
 	cfg := nodebuilder.DefaultConfig(node.Bridge)
-	const defaultTimeInterval = time.Second * 10
 	setTimeInterval(cfg, defaultTimeInterval)
-
 	bridge := sw.NewNodeWithConfig(node.Bridge, cfg)
-
-	ctx, cancel := context.WithTimeout(context.Background(), swamp.DefaultTestTimeout)
-	t.Cleanup(cancel)
-
 	err := bridge.Start(ctx)
 	require.NoError(t, err)
-	bridgeAddr := host.InfoFromHost(bridge.Host)
 
+	// use BN as the bootstrapper
+	bootstrapper := host.InfoFromHost(bridge.Host)
+
+	// create FN with BN as bootstrapper
 	cfg = nodebuilder.DefaultConfig(node.Full)
 	setTimeInterval(cfg, defaultTimeInterval)
 	full := sw.NewNodeWithConfig(
 		node.Full,
 		cfg,
-		nodebuilder.WithBootstrappers([]peer.AddrInfo{*bridgeAddr}),
+		nodebuilder.WithBootstrappers([]peer.AddrInfo{*bootstrapper}),
 	)
 
+	// create LN with BN as bootstrapper
 	cfg = nodebuilder.DefaultConfig(node.Light)
 	setTimeInterval(cfg, defaultTimeInterval)
-	cfg.P2P.PeerExchange = true
 	light := sw.NewNodeWithConfig(
 		node.Light,
 		cfg,
-		nodebuilder.WithBootstrappers([]peer.AddrInfo{*bridgeAddr}),
+		nodebuilder.WithBootstrappers([]peer.AddrInfo{*bootstrapper}),
 	)
+
+	// start FN and LN and ensure they are both connected to BN as a bootstrapper
 	nodes := []*nodebuilder.Node{full, light}
-	ch := make(chan struct{})
-	sub, err := light.Host.EventBus().Subscribe(&event.EvtPeerConnectednessChanged{})
-	require.NoError(t, err)
-	defer sub.Close()
 	for index := range nodes {
 		require.NoError(t, nodes[index].Start(ctx))
-		assert.Equal(t, *bridgeAddr, nodes[index].Bootstrappers[0])
-		assert.True(t, nodes[index].Host.Network().Connectedness(bridgeAddr.ID) == network.Connected)
+		assert.Equal(t, *bootstrapper, nodes[index].Bootstrappers[0])
+		assert.True(t, nodes[index].Host.Network().Connectedness(bootstrapper.ID) == network.Connected)
 	}
-	addrFull := host.InfoFromHost(full.Host)
-	go func() {
-		for e := range sub.Out() {
-			connStatus := e.(event.EvtPeerConnectednessChanged)
-			if connStatus.Peer == full.Host.ID() && connStatus.Connectedness == network.NotConnected {
-				ch <- struct{}{}
-			}
+
+	for {
+		if ctx.Err() != nil {
+			t.Fatal(ctx.Err())
+		}
+		if light.Host.Network().Connectedness(host.InfoFromHost(full.Host).ID) == network.Connected {
+			// LN discovered FN successfully and is now connected
+			break
 		}
```

Comment:
Polling loop without delay would eat cpu in tests and slow down CI. Why not stick with event bus as it used to be?

Reply:
listening for a connected event does not work for some reason even though it definitely should. Will continue investigating.

Reply:
Probably because you need to spin out listener before even happens. With for loop you check after the event

Reply:
I started an event listening loop before the event happens and it still doesn't work. Will continue investigating but if this is blocking, we can move on other cleanussy PRs.

Reply:
Investigate why subscription isn't working in this test.

Reply:
https://github.com/celestiaorg/celestia-node/issues/2649